crates/hdx_ast/src/css/values/color_hdr/types.rs

```diff
@@ -1,2 +1,82 @@
-pub(crate) use crate::css::types::*;
-pub(crate) use crate::css::units::*;
+use bumpalo::collections::Vec;
+use hdx_lexer::Cursor;
+use hdx_parser::{diagnostics, CursorStream, Is, Parse, Parser, Peek, Result as ParserResult, ToCursors, T};
+
+mod func {
+	use hdx_parser::custom_function;
+	custom_function!(DynamicRangeLimitMix, atom!("dynamic-range-limit-mix"));
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize), serde())]
+pub struct DynamicRangeLimitMix<'a> {
+	function: T![Function],
+	values: Vec<'a, (T![Ident], T![Dimension::%], Option<T![,]>)>,
+	close: Option<T![')']>,
+}
+
+impl<'a> Peek<'a> for DynamicRangeLimitMix<'a> {
+	fn peek(p: &Parser<'a>) -> bool {
+		p.peek::<func::DynamicRangeLimitMix>()
+	}
+}
+
+impl<'a> Parse<'a> for DynamicRangeLimitMix<'a> {
+	fn parse(p: &mut Parser<'a>) -> ParserResult<Self> {
+		let function = p.parse::<T![Function]>()?;
+		let c = function.into();
+		if !func::DynamicRangeLimitMix::is(p, c) {
+			Err(diagnostics::Unexpected(c.into(), c.into()))?;
+		}
+		let mut values = Vec::new_in(p.bump());
+		loop {
+			if p.at_end() {
+				return Ok(Self { function, values, close: None });
+			}
+			if p.peek::<T![')']>() {
+				return Ok(Self { function, values, close: Some(p.parse::<T![')']>()?) });
+			}
+			let ident = p.parse::<T![Ident]>()?;
+			let length = p.parse::<T![Dimension::%]>()?;
+			let c: Cursor = length.into();
+			if !(0.0..=100.0).contains(&c.token().value()) {
+				Err(diagnostics::NumberOutOfBounds(c.token().value(), format!("{:?}", 0.0..=100.0), c.into()))?
+			}
+			let comma = p.parse_if_peek::<T![,]>()?;
+			values.push((ident, length, comma));
+		}
+	}
+}
+
+impl<'a> ToCursors<'a> for DynamicRangeLimitMix<'a> {
```

Comment:
Every AST node must also implement the `ToCursors` trait which is given a `CursorStream` and must feed its cursors, in order, to the cursor stream. This then lets us write the cursor stream back out to a string.